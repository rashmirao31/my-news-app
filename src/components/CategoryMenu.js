import React from 'react'
import { Menu } from 'semantic-ui-react';

class CategoryMenu extends React.Component {

  	constructor(props) {
	    super(props);
	    this.state = {
	      activeItem: this.props.selectedCategory,
	      searchText: ''
	    };
	    this.handleItemClick = this.handleItemClick.bind(this);
	    this.onSearchTextChange = this.onSearchTextChange.bind(this);
	    this.onKeyPress = this.onKeyPress.bind(this);
	    this.handleSearchButtonClick = this.handleSearchButtonClick.bind(this);
  	};

  	componentDidUpdate(prevProps, prevState) {
  		if (prevProps.selectedCategory !== this.props.selectedCategory) {
  			this.setState({
	  			activeItem: this.props.selectedCategory,
	  			searchText: ''
	  		});
  		}  		
  	};

  	handleItemClick(event, {name}) {
	    this.setState({
	      activeItem: name,
	      searchText: ''
	    });
	    this.props.updateSelectedCategory(name);
  	};

  	handleSearchButtonClick(event) {
  		this.setState({
	      activeItem: ''
	    });
	    this.props.updateSelectedCategory('');
	    this.props.updateSearchText(this.state.searchText);
  	};

  	onSearchTextChange(event) {
  		this.setState({
	      searchText: event.target.value
	    });
  	};

  	onKeyPress(event) {
  		var enterKeyCode = 13;
  		var keyCodeEntered = event.keyCode || event.which; 		

	    if (keyCodeEntered === enterKeyCode) {
  			this.handleSearchButtonClick();
  		}
  	};

  	render() {
    	const { activeItem } = this.state;

    	return (
	      	<div className='ui container' data-testid='category-menu'>
		        <Menu pointing>
		          <Menu.Item
		            name='general'
		            active={activeItem === 'general'}
		            onClick={this.handleItemClick}
		          />
		          <Menu.Item
		            name='health'
		            active={activeItem === 'health'}
		            onClick={this.handleItemClick}
		          />
		          <Menu.Item
		            name='business'
		            active={activeItem === 'business'}
		            onClick={this.handleItemClick}
		          />
		          <Menu.Item
		            name='technology'
		            active={activeItem === 'technology'}
		            onClick={this.handleItemClick}
		          />
		          <Menu.Item
		            name='sports'
		            active={activeItem === 'sports'}
		            onClick={this.handleItemClick}
		          />
		          <Menu.Item
		            name='entertainment'
		            active={activeItem === 'entertainment'}
		            onClick={this.handleItemClick}
		          />
		          <Menu.Item
		            name='science'
		            active={activeItem === 'science'}
		            onClick={this.handleItemClick}
		          />
		          <Menu.Menu position='right'>
		            <Menu.Item>		              
		            	<input type='text' 
		            		data-testid='search-input'
		            		placeholder={window.defaultAppSettings.searchInputPlaceholder} 
		            		value={this.state.searchText} 
		            		onKeyPress={this.onKeyPress} 
		            		onChange={this.onSearchTextChange}
		            	/>
		            	<button onClick={this.handleSearchButtonClick}
		            		data-testid='search-button'>
		            		{window.defaultAppSettings.searchButtonText}
		            	</button>
		            </Menu.Item>
		          </Menu.Menu>
		        </Menu>
	      	</div>
	    );
  	};
  
}

export default CategoryMenu