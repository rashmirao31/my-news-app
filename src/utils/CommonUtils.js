var CommonUtils = {

	getConfigList: function() {
		let newsApiUrl = "https://newsapi.org/v2/";
		return {
			"newsApiUrl": newsApiUrl,
			"newsApiTopHeadlinesUrl": newsApiUrl + "top-headlines",
			"newsApiEverythingUrl": newsApiUrl + "everything",
			"apiKey": "687cf8fdc1d7435eb10214b70c78788d",
			"countryKey": "gb",
			"countryName": "United Kingdom",
			"category": "general",
			"timezoneKey": "Europe/London",
			"countryOptions": [
			  { key: 'gb', value: 'gb', flag: 'gb', text: 'United Kingdom' },
			  { key: 'us', value: 'us', flag: 'us', text: 'USA' },
			  { key: 'in', value: 'in', flag: 'in', text: 'India' }
			],
			"countryMenuTitle": "Country:",
			"countryMenuPlaceholder": "Select Country",
			"headerTimeLabel": "Updated at: ",
			"pageTitleSuffixLabel": "News",
			"searchInputPlaceholder": "Search...",
			"searchButtonText": "Search",
			"topStoriesText": "Top Stories",
			"searchResultHeaderText": "Search Results For",
			"loadingText": "Please wait...",
			"noDataText": "No results found",
			"readMoreText": "Read more",
			"errorScenarioText": "Something went wrong. Please try again after sometime."
		};
	},


};

export default CommonUtils;