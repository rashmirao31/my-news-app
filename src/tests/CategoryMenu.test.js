import React from 'react';
import ReactDOM from 'react-dom';
import CommonUtils from './../utils/CommonUtils';
import CategoryMenu from './../components/CategoryMenu.js';

import { render, fireEvent, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from 'react-test-renderer';

beforeEach(() => {
  window.defaultAppSettings = CommonUtils.getConfigList();
  return;
});

afterEach(cleanup);

it('renders without crashing', () => {
 	const div = document.createElement('div');
	ReactDOM.render(<CategoryMenu />, div);
});

it('matches snapshot', () => {
 	const tree = renderer.create(
    	<CategoryMenu/>
	).toJSON();
 	expect(tree).toMatchSnapshot();
});

it('renders with the default category', () => {
 	const { getByText } = render(
    	<CategoryMenu />
	);
	let defaultCategory = window.defaultAppSettings.category;
	defaultCategory = defaultCategory.substring(0,1).toUpperCase() + defaultCategory.substring(1);
	expect(getByText(defaultCategory)).toBeVisible();	
});

it('renders with the search button', () => {
 	const { getByText } = render(
    	<CategoryMenu />
	);
	let searchButtonText = window.defaultAppSettings.searchButtonText;
	expect(getByText(searchButtonText)).toBeVisible();
});

it('renders with the input textbox', () => {
 	const { getByTestId } = render(
    	<CategoryMenu />
	);
	expect(getByTestId("search-input")).toBeVisible();
});

it('renders with the selected category', () => {
 	const selectedCategory = "entertainment";
 	const component = renderer.create(
    	<CategoryMenu selectedCategory={selectedCategory}/>
	);
	const instance = component.getInstance();
  	expect(instance.state.activeItem).toBe(selectedCategory);
});

it('re-renders with selected category', () => {
	const selectedCategory = "health";
	let selectedCategoryName = selectedCategory.substring(0,1).toUpperCase() + selectedCategory.substring(1);
	const updateSelectedCategoryMock = jest.fn();
	const updateSearchTextMock = jest.fn();
	const { getByText } = render(<CategoryMenu 
		selectedCategory={window.defaultAppSettings.category}  
		updateSelectedCategory = {updateSelectedCategoryMock}
	/>);

	fireEvent.click(getByText(selectedCategoryName));
	expect(updateSelectedCategoryMock).toHaveBeenCalledWith(selectedCategory);
});

it('re-renders by passing the text searched to its parent', () => {
 	const updateSelectedCategoryMock = jest.fn();
 	const updateSearchTextMock = jest.fn();
 	const searchText = "God";
 	const { getByTestId } = render(<CategoryMenu 
		selectedCategory={window.defaultAppSettings.category}
		updateSelectedCategory={updateSelectedCategoryMock}
		updateSearchText={updateSearchTextMock}
	/>);
	fireEvent.change(getByTestId("search-input"), { target : { value: searchText}});
	fireEvent.click(getByTestId("search-button"));
  	expect(updateSearchTextMock).toHaveBeenCalledWith(searchText);
});

it('removes category selection when anything is searched', () => {
 	const updateSelectedCategoryMock = jest.fn();
 	const updateSearchTextMock = jest.fn();
 	const { getByTestId } = render(<CategoryMenu 
		selectedCategory={window.defaultAppSettings.category}
		updateSelectedCategory={updateSelectedCategoryMock}
		updateSearchText={updateSearchTextMock}
	/>);
	fireEvent.click(getByTestId("search-button"));
  	expect(updateSelectedCategoryMock).toHaveBeenCalledWith('');
});