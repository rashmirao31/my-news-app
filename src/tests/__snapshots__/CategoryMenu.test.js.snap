// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="ui container"
  data-testid="category-menu"
>
  <div
    className="ui pointing menu"
  >
    <a
      className="item"
      onClick={[Function]}
    >
      General
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Health
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Business
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Technology
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Sports
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Entertainment
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Science
    </a>
    <div
      className="right menu"
    >
      <div
        className="item"
        onClick={[Function]}
      >
        <input
          data-testid="search-input"
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="Search..."
          type="text"
          value=""
        />
        <button
          data-testid="search-button"
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </div>
  </div>
</div>
`;
